crates/ra_ide_api/src/impls.rs

```diff
@@ -0,0 +1,74 @@
+use ra_db::{SourceDatabase};
+use ra_syntax::{
+    AstNode, ast,
+    algo::find_node_at_offset,
+};
+use hir::{db::HirDatabase, source_binder};
+use test_utils::tested_by;
+
+use crate::{FilePosition, NavigationTarget, db::RootDatabase, RangeInfo};
+
+pub(crate) fn goto_implementation(
+    db: &RootDatabase,
+    position: FilePosition,
+) -> Option<RangeInfo<Vec<NavigationTarget>>> {
+    let file = db.parse(position.file_id);
+    let syntax = file.syntax();
+
+    let krate_id = db.crate_for(position.file_id).pop()?;
+    let krate = hir::Crate { crate_id: krate_id };
+    let module = source_binder::module_from_position(db, position)?;
+
+    let (range, ty) = if let Some(struct_def) =
+        find_node_at_offset::<ast::StructDef>(syntax, position.offset)
+    {
+        (
+            struct_def.syntax().range(),
+            source_binder::struct_from_module(db, module, &struct_def).ty(db),
+        )
+    } else if let Some(enum_def) = find_node_at_offset::<ast::EnumDef>(syntax, position.offset) {
+        (
+            enum_def.syntax().range(),
+            source_binder::enum_from_module(db, module, &enum_def).ty(db),
+        )
+    } else {
+        return None;
+    };
+
+    let impls = db.impls_in_crate(krate);
+
+    let navs = impls
+        .lookup_impl_blocks(db, &ty)
+        .map(|it| module.impl_source(db, it.id()))
```

Comment:
Note that this might be the wrong module: struct can be in `a.rs`, impl can be in `b.rs`.